mysql-test/suite/innodb/t/innodb-lock-inherit-read_commited.test

```diff
@@ -0,0 +1,114 @@
+--echo #
+--echo # Bug #21025880	DUPLICATE UK VALUES IN READ-COMMITTED(AGAIN)
+--echo #
+
+--source include/have_innodb.inc
+--source include/have_debug.inc
+--source include/have_debug_sync.inc
+
+SET @save_frequency = @@GLOBAL.innodb_purge_rseg_truncate_frequency;
+SET GLOBAL innodb_purge_rseg_truncate_frequency=1;
+
+let $i=0;
+
+while ($i <=1 )
+{
+
+CREATE TABLE t1 (
+       a INT NOT NULL,
+       b INT NOT NULL,
+       PRIMARY KEY(b),
+       UNIQUE KEY(a))
+ENGINE=INNODB;
+
+SET @old_innodb_stats_auto_recalc = @@innodb_stats_auto_recalc;
+SET GLOBAL innodb_stats_auto_recalc = OFF;
+
+# Block purge
+connect  purge_control,localhost,root;
+START TRANSACTION WITH CONSISTENT SNAPSHOT;
+connection default;
+
+SET @old_tx_isolation = @@tx_isolation;
+SET GLOBAL tx_isolation = 'READ-COMMITTED';
+
+SET @old_innodb_lock_wait_timeout = @@innodb_lock_wait_timeout;
+SET GLOBAL innodb_lock_wait_timeout = 1;
+
+--connect(con1,localhost,root,,)
+
+# Create and delete-mark an index record
+
+INSERT INTO t1 VALUES (1,1),(2,2);
+DELETE FROM t1;
+
+SET debug_sync = 'row_ins_sec_index_entry_dup_locks_created SIGNAL
+con1_locks_done WAIT_FOR con1_go';
+SET debug_sync = 'ha_commit_trans_after_acquire_commit_lock SIGNAL
+con1_insert_done WAIT_FOR con1_finish';
+--send
+
+if ($i == 0)
+{
+REPLACE INTO t1 VALUES (1,2);
+}
+
+if ( $i == 1)
+{
+INSERT INTO t1 values (1,2) ON DUPLICATE KEY UPDATE a=2;
+}
+--connect(con2,localhost,root,,)
+
+SET debug_sync = 'now WAIT_FOR con1_locks_done';
+
+SET debug_sync = 'lock_wait_suspend_thread_enter SIGNAL con2_blocked
+WAIT_FOR con2_go';
+SET debug_sync = 'ha_commit_trans_after_acquire_commit_lock SIGNAL
+con2_insert_done WAIT_FOR con2_finish';
+SET debug_sync = 'ib_after_row_insert SIGNAL con2_insert_done';
+
+--send
+REPLACE INTO t1 VALUES (1,3);
+
+--connection default
+SET debug_sync = 'now WAIT_FOR con2_blocked';
+
+connection purge_control;
+COMMIT;
+disconnect purge_control;
+connection default;
+
+# Wait for purge to delete the delete-marked record
+--source ../../innodb/include/wait_all_purged.inc
+
+SET debug_sync = 'now SIGNAL con2_go WAIT_FOR con2_insert_done';
+SET debug_sync = 'now SIGNAL con1_go WAIT_FOR con1_insert_done';
+
+SET debug_sync = 'now SIGNAL con1_finish';
+SET debug_sync = 'now SIGNAL con2_finish';
+
+--connection con1
+--reap
```

Comment:
This test occasionally fails on several builders. I will have to check this deeper. It has failed in both iterations of the loop:
```diff
@@ -46,6 +46,8 @@
 SET debug_sync = 'now SIGNAL con1_finish';
 SET debug_sync = 'now SIGNAL con2_finish';
 connection con1;
+Warnings:
+Warning	1639	debug sync point wait timed out
 disconnect con1;
 connection con2;
 disconnect con2;
```
```diff
@@ -104,6 +104,8 @@
 SET debug_sync = 'now SIGNAL con1_finish';
 SET debug_sync = 'now SIGNAL con2_finish';
 connection con1;
+Warnings:
+Warning	1639	debug sync point wait timed out
 disconnect con1;
 connection con2;
 disconnect con2;
```

Reply:
I was able to repeat the `DEBUG_SYNC` hang on my system when running 17×10 tests. After I applied the following cleanup, the test did not hang on my system:
```diff
diff --git a/mysql-test/suite/innodb/t/innodb-lock-inherit-read_commited.test b/mysql-test/suite/innodb/t/innodb-lock-inherit-read_commited.test
index 1c887911069..d9c933fa9b5 100644
--- a/mysql-test/suite/innodb/t/innodb-lock-inherit-read_commited.test
+++ b/mysql-test/suite/innodb/t/innodb-lock-inherit-read_commited.test
@@ -63,8 +63,8 @@ SET debug_sync = 'now WAIT_FOR con1_locks_done';
 
 SET debug_sync = 'lock_wait_suspend_thread_enter SIGNAL con2_blocked
 WAIT_FOR con2_go';
-SET debug_sync = 'ha_commit_trans_after_acquire_commit_lock SIGNAL
-con2_insert_done WAIT_FOR con2_finish';
+SET debug_sync = 'ha_commit_trans_after_acquire_commit_lock
+WAIT_FOR con2_finish';
 SET debug_sync = 'ib_after_row_insert SIGNAL con2_insert_done';
 
 --send
@@ -85,11 +85,12 @@ SET debug_sync = 'now SIGNAL con2_go WAIT_FOR con2_insert_done';
 SET debug_sync = 'now SIGNAL con1_go WAIT_FOR con1_insert_done';
 
 SET debug_sync = 'now SIGNAL con1_finish';
-SET debug_sync = 'now SIGNAL con2_finish';
 
 --connection con1
 --reap
 --disconnect con1
+--connection default
+SET debug_sync = 'now SIGNAL con2_finish';
 
 --connection con2
 --error 0,ER_LOCK_WAIT_TIMEOUT
```
Only one other added test `innodb.innodb-ac-non-locking-select` makes use of `DEBUG_SYNC`, in a much simpler pattern. I successfully ran 68×30=2040 repetitions of it on my system.